File: src/main/java/com/matera/restserver/controller/EmployeeController.java
package com.matera.restserver.controller;

import java.io.IOException;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.server.ResponseStatusException;
import org.springframework.web.util.UriComponents;
import org.springframework.web.util.UriComponentsBuilder;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.matera.restserver.dto.EmployeeDTO;
import com.matera.restserver.exception.EntityExistsException;
import com.matera.restserver.exception.EntityNotFoundException;
import com.matera.restserver.service.EmployeeService;

@RestController
@RequestMapping("/employee")
public class EmployeeController {

	private static final String BAD_PARSING = "Something went wrong while parsing the body of the request :(";

	@Autowired
	private EmployeeService service;

	@Autowired
	private ObjectMapper mapper;

	private EmployeeDTO employee;

	/**
	 * Creates an employee.
	 * 
	 * @param reqBody A JSON that represents the employee being created, which will
	 *                come in the request body.
	 * @return the employee created
	 */
	@PostMapping(produces = "application/json", consumes = "application/json")
	public ResponseEntity<EmployeeDTO> createEmployee(@RequestBody String reqBody, UriComponentsBuilder uri) {
		try {
			employee = mapper.readValue(reqBody, EmployeeDTO.class);
		} catch (IOException e1) {
			throw new ResponseStatusException(HttpStatus.BAD_REQUEST, BAD_PARSING, e1);
		}
		try {
			employee.setId(service.create(employee));
		} catch (EntityExistsException e) {
			throw new ResponseStatusException(HttpStatus.CONFLICT, e.getMessage(), e);
		}

		// Return the location on the header
		UriComponents uriComponents = uri.path("/employee/{id}").buildAndExpand(employee.getId());

		return ResponseEntity.created(uriComponents.toUri()).body(employee);
	}

	/**
	 * Searchs all active employees.
	 * 
	 * @return All employees on the database.
	 */
	@GetMapping(produces = "application/json")
	public List<EmployeeDTO> findAllEmployees() {
		try {
			return service.findAll();
		} catch (EntityNotFoundException e) {
			throw new ResponseStatusException(HttpStatus.NOT_FOUND, e.getMessage(), e);
		}
	}

	/**
	 * Searchs for an active employee with given id.
	 * 
	 * @param id id of the employee.
	 * @return employee with given id, if it exists.
	 */
	@GetMapping(value = "/{id}", produces = "application/json")
	public EmployeeDTO findEmployee(@PathVariable Long id) {
		try {
			return service.find(id);
		} catch (EntityNotFoundException e) {
			throw new ResponseStatusException(HttpStatus.NOT_FOUND, e.getMessage(), e);
		}
	}

	/**
	 * Deletes (deactivates) an employee with given id.
	 * 
	 * @param id
	 * @return the employee deactivated
	 */
	@DeleteMapping(value = "/{id}", produces = "application/json")
	public EmployeeDTO deleteEmployee(@PathVariable Long id) {
		try {
			return service.delete(id);
		} catch (EntityNotFoundException e) {
			throw new ResponseStatusException(HttpStatus.NOT_FOUND, e.getMessage(), e);

		}

	}

	/**
	 * Updates an employee with given id.
	 * 
	 * @param id
	 * @return the employee updated
	 */
	@PutMapping(value = "/{id}", produces = "application/json", consumes = "application/json")
	public EmployeeDTO updateEmployee(@PathVariable Long id, @RequestBody String reqBody) {
		try {
			employee = mapper.readValue(reqBody, EmployeeDTO.class);
		} catch (IOException e1) {
			throw new ResponseStatusException(HttpStatus.BAD_REQUEST, BAD_PARSING, e1);
		}

		// Ignore ID if informed on the request body
		employee.setId(id);

		try {
			service.update(employee);
		} catch (EntityNotFoundException e) {
			throw new ResponseStatusException(HttpStatus.BAD_REQUEST, e.getMessage(), e);
		}

		return employee;
	}

}

File: src/main/java/com/matera/restserver/controller/EmployeeController.kt
package com.matera.restserver.controller

import com.fasterxml.jackson.databind.ObjectMapper
import com.matera.restserver.dto.EmployeeDTO
import com.matera.restserver.exception.EntityExistsException
import com.matera.restserver.exception.EntityNotFoundException
import com.matera.restserver.service.EmployeeService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException
import org.springframework.web.util.UriComponentsBuilder
import java.io.IOException

@RestController
@RequestMapping("/employee")
class EmployeeController {
    @Autowired
    private val service: EmployeeService? = null

    @PostMapping(produces = ["application/json"], consumes = ["application/json"])
    fun createEmployee(@RequestBody employee: EmployeeDTO?, uri: UriComponentsBuilder): ResponseEntity<EmployeeDTO?> {
        try {
            employee?.id = service!!.create(employee)
        } catch (e: EntityExistsException) {
            throw ResponseStatusException(HttpStatus.CONFLICT, e.message, e)
        }

        // Return the location on the header
        val uriComponents = uri.path("/employee/{id}").buildAndExpand(employee!!.id)
        return ResponseEntity.created(uriComponents.toUri()).body(employee)
    }

    @GetMapping(produces = ["application/json"])
    fun findAllEmployees(): List<EmployeeDTO> {
        return try {
            service!!.findAll()
        } catch (e: EntityNotFoundException) {
            throw ResponseStatusException(HttpStatus.NOT_FOUND, e.message, e)
        }
    }

    @GetMapping(value = ["/{id}"], produces = ["application/json"])
    fun findEmployee(@PathVariable id: Long?): EmployeeDTO {
        return try {
            service!!.find(id)
        } catch (e: EntityNotFoundException) {
            throw ResponseStatusException(HttpStatus.NOT_FOUND, e.message, e)
        }
    }

    @DeleteMapping(value = ["/{id}"], produces = ["application/json"])
    fun deleteEmployee(@PathVariable id: Long?): EmployeeDTO {
        return try {
            service!!.delete(id)
        } catch (e: EntityNotFoundException) {
            throw ResponseStatusException(HttpStatus.NOT_FOUND, e.message, e)
        }
    }

    @PutMapping(value = ["/{id}"], produces = ["application/json"], consumes = ["application/json"])
    fun updateEmployee(@PathVariable id: Long?, @RequestBody employee: EmployeeDTO?): EmployeeDTO? {
        // Ignore ID if informed on the request body
        employee?.id = id
        try {
            service!!.update(employee)
        } catch (e: EntityNotFoundException) {
            throw ResponseStatusException(HttpStatus.BAD_REQUEST, e.message, e)
        }
        return employee
    }

    companion object {
        private const val BAD_PARSING = "Something went wrong while parsing the body of the request :("
    }
}